openeo_pg_parser_networkx/resolving_utils.py

```diff
@@ -0,0 +1,217 @@
+    like apply and reduce.
+
+    Parameters:
+        process_graph (dict[str,Any]): The Process Graph which is to be resolved.
+        process_registry (ProcessRegistry): The process registry containing all predefined processes,
+                                            and optionally all relevant UDPs in the 'user' namespace
+        get_udp_spec (Callable[[str], dict]): A callable which returns the spec of a given udp, takes process_id as a parameter.
+                                              It being a "spec" just requires it to be a dictionary containing the key 'process_graph' which
+                                              contains the relevant Process-Graph as its value.
+
+    Returns:
+        resolved_process_graph (dict[str,Any]): The resolved process graph.
+    '''
+    process_graph = _unpack_process_graph(
+        process_graph=process_graph,
+        process_registry=process_registry,
+        get_udp_spec=get_udp_spec,
+    )
+    process_graph = _resolve_sub_process_graphs(
+        process_graph=process_graph,
+        process_registry=process_registry,
+        get_udp_spec=get_udp_spec,
+    )
+    return process_graph
+
+
+def _unpack_process_graph(
+    process_graph: dict[str, Any],
+    process_registry: ProcessRegistry,
+    get_udp_spec: Callable[[str], dict] = None,
+):
+    '''
+    Wraps method calls necessary for resolving a process graph without sub-graphs within
+    nodes like apply and reduce.
+    '''
+    _fill_in_processes(process_graph, process_registry, get_udp_spec)
+    root_result = _adjust_references(process_graph)
+    process_graph = _flatten_graph(process_graph)
+    _remove_non_root_result(process_graph, root_result)
+    return process_graph
+
+
+'''
+Finds and resolves any sub process graphs within nodes like apply and reduce.
+'''
+
+
+def _resolve_sub_process_graphs(
+    process_graph: dict,
+    process_registry: ProcessRegistry,
+    get_udp_spec: Callable[[str], dict] = None,
+):
+    for _, node in process_graph.items():
+        if 'process' in node['arguments']:
+            node['arguments']['process']['process_graph'] = resolve_process_graph(
+                node['arguments']['process']['process_graph'],
+                process_registry,
+                get_udp_spec,
+            )
+    return process_graph
+
+
+def _fill_in_processes(
+    process_graph: dict[str, Any],
+    process_registry: ProcessRegistry,
+    get_udp_spec: Callable[[str], dict] = None,
+):
+    """
+    Recursively fill in UDPs with their respective definitions until
+    only predefined processes remain.
+
+    Ignores subgraphs, those get resolved individually.
+    """
+    for process_replacement_id, process in process_graph.items():
+        process_id = process['process_id']
+
+        if ('predefined', process_id) not in process_registry:
+            if ('user', process_id) not in process_registry and get_udp_spec is not None:
```

Comment:
We don't have access to the user_id in the resolving logic, but I could look for the process_id across all namespaces I guess.

I just figured that having it directly accessible would be preferable, but I can change it.

Reply:
Ok, let's look in all namespaces then - thanks!

Reply:
Can we assume the uniqueness of the process_id across namespaces? We let users fill up the process_registry themselves after all.

Reply:
Hmm, don't think we can assume this - actually, is there any rule that prevents users from naming their UDP like a predefined process?

Tbh, as long as we guarantee that predefined processes always get preferred to UDPs if there's a collision, that should be fine.

Reply:
I meant this more as in multiple UDPs with the same name across different namespaces that get passed into the process_registry out of carelessness. Predefined processes are already preferred over UDPs because if a process is found in the DEFAULT_NAMESPACE it isn't resolved at all (even if it is empty and the user intended for it to be a UDP).

I mean, forcing a 'user' namespace only fixes this problem by forcing users to have a single clear UDP which is guaranteed to be the intended one, but also adds more confusion, because now a special process_registry has to be prepared if this functionality is to be used.

I don't know if these kinds of collisions would ever come up, but maybe we should just disallow backends to pass an already filled process_registry and only allow usage with the callback function. That way we always get a single, clear UDP implementation back and we don't have to think about any naming clashes.

Reply:
I changed this to just using an optional argument "namespace" which defaults to user and gets passed on to the get_udp_spec call.